### ISSUE TITLE
Improve implicit .NET standard facade references

### ISSUE THREAD

**Issue by jnm2** [dotnet/sdk#1957]
If you reference [System.Collections.Immutable 1.4.0](https://www.nuget.org/packages/System.Collections.Immutable/1.4.0) in a net462 project, it picks the netstandard2.0 System.Collections.Immutable dll. Unfortunately, the build system copies not just `netstandard.dll` but all the .NET Standard 1.\* facade assemblies as well. This means 72 unreferenced, unused System.\* and Microsoft.\* DLLs show up for deployment. This is decidedly unpleasant.

Would you please make the SDK smart enough to only add the netstandard1.\* assemblies if there is at least one reference to a netstandard1.\* assembly?
And, independently, add `netstandard.dll` only if there is at least one reference to a netstandard2.\* assembly?

Here is the target which currently lumps both 1.\* and 2.\* in at once or none at all:

https://github.com/dotnet/sdk/blob/83ba345017d368c80e2d24090a2507a0f4627001/src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets#L90-L95

This has been the only reason I've backed away from using System.Collections.Immutable in many projects.

**Comment by livarcocc** [comment]
cc @dsplaisted @ericstj

**Comment by ericstj** [comment]
You'd actually need a more assemblies than just netstandard.dll, since netstandard.dll requires more API than  was included inbox.  The following assemblies have types that are part of netstandard2 but not in .net 4.6.2.  /cc @weshaggard @joperezr 

- System.Data.Common
- System.Diagnostics.StackTrace
- System.Diagnostics.Tracing
- System.Globalization.Extensions
- System.IO.Compression
- System.Net.Http
- System.Net.Sockets
- System.Runtime.InteropServices.RuntimeInformation
- System.Security.Cryptography.Algorithms
- System.Threading.Overlapped
- System.ValueTuple
- System.Xml.XPath.XDocument

@jnm2 another thing you can look at, if you own the application is to try out the [dependency trimmer](https://github.com/dotnet/standard/blob/master/Microsoft.Packaging.Tools.Trimming/docs/trimming.md).

**Comment by jnm2** [comment]
Funny you mention that. We've been using a separate dependency trimmer which requires you to explicitly opt in or out of each assembly not transitively referenced from the main build output. I wasn't looking forward to adding 72 lines to the csproj and it seems there would be more general interest in not outputting all of them.

I may try your dependency trimmer though. Even better if it detects, for example, that none of the referenced .NET Standard libraries contains code which would load `System.Security.Cryptography.Algorithms` and therefore skips it. Is that capability tracked?

Regardless, I do believe that it would be in the general interest to copy only those 13 assemblies rather than all 96 when you reference a netstandard2.\* assembly but not a netstandard1.\* one.

**Comment by ericstj** [comment]
> Even better if it detects, for example, that none of the referenced .NET Standard libraries contains code which would load System.Security.Cryptography.Algorithms and therefore skips it.

Yes, we drill through type-forwards.

**Comment by jnm2** [comment]
Is https://github.com/dotnet/standard/issues/667 the right place to file issues with your dependency trimmer?

**Comment by ericstj** [comment]
Yep, that's correct.  In the future you can tag me.  Sorry I didn't see it.

**Comment by jnm2** [comment]
No problem. Thanks for getting me unblocked with dotnet/standard#667.


With `<TrimUnusedDependencies>true</TrimUnusedDependencies>`, `<RootPackageReference>false</RootPackageReference>`, and `<PackageReference Include="Microsoft.Packaging.Tools.Trimming" Version="1.1.0-preview1-25818-01" />`, no dependencies are being trimmed when building from Visual Studio or `msbuild /t:Rebuild`. I believe I'm following https://github.com/dotnet/standard/blob/master/Microsoft.Packaging.Tools.Trimming/docs/trimming.md#how-to-use. Did I miss something else?

**Comment by ericstj** [comment]
It's working for me.  Here's what I see on a simple project that uses one NETStandard library, but doesn't need any of the 1.x facades:
```
TrimFilesOnBuild:
  Trimmed 95 out of 98 files for a savings of 2.47 MB
  Final app size is 0.27 MB
```
You may want to delete your output directory to get in a clean state.  It's possible incremental clean lost track of the files.

If that doesn't work can you share a repro project?

**Comment by jnm2** [comment]
I fixed a typo and it started working. 🤦‍♂️ All the right assemblies are now trimmed! [New issue](https://github.com/dotnet/standard/issues/670) blocking me that looks legit.

**Comment by livarcocc** [comment]
Closing as the issue seem to have been addressed and new issues were filed against dotnet/standard.
